module2/lessons/serializers.md

```diff
@@ -0,0 +1,218 @@
+
+You can confirm this worked by opening your rails console and taking a look at the contents of your development database.
+
+Start up your server using `rails server` and you're ready to go!
+
+## Responses
+
+Use Postman to view the current responses that your API is providing to the routes listed below:
+
+- api/v1/songs
+- api/v1/songs/:id
+
+So we have our responses from our server, but it isn’t JSON API 1.0 And it has this created at and updated at stuff which we don’t want. So what do we do? We need to use a serializer.
+
+## Customizing JSON
+
+This is some practice time for you. 
+
+1. Create a serializer for `Song` and build out a hash that will look like the following *WITHOUT THE USE OF A GEM.* (You only need to do this for the show)
+
+```ruby
+{
+  "data": [
+    {
+      "id": "1",
+      "title": "Legend Has It",
+      "length": 2301
+    }
+  ]
+}
+```
+
+That was a pain in the butt, wasn’t it? Creating serializers by hand that are JSON API 1.0 for everything we want to make an API for can certainly be time consuming. There are better ways.
+
+## Using the jsonapi-serializer gem
+
+You can view the docs on the jsonapi-serializer gem [here](https://github.com/jsonapi-serializer/jsonapi-serializer#installation).
+
+Add it to your Gemfile
+
+```ruby
+gem "jsonapi-serializer"
+```
+
+And go ahead and
+
+```bash
+$ bundle install
+```
+
+This gem gives us a built in generator to make ourselves all of the serializers we could possibly want.
+
+```bash
+$ rails g serializer Song title
+```
+
+We can take a look and see what’s inside it.
+
+*app/serializers/store_serializer.rb*
+
+```ruby
+class SongSerializer
+  include JSONAPI::Serializer
+  attributes :title
+end
+```
+
+Now that we have our serializer, we need to edit our controller to use said serializer.
+
+*app/controllers/api/v1/songs_controller.rb*
+
+```ruby
+class Api::V1::SongsController < ApplicationController
+  def index
+    render json: SongSerializer.new(Song.all)
+  end
+
+  def show
+    render json: SongSerializer.new(Song.find(params[:id]))
+  end
+end
+```
+
+We can see that instead of just letting our ActiveRecord be rendered in JSON, we are going to take our collections, and send them to the serializer, and have the result of THAT be converted to JSON.
+
+This Gem can also be used to easily set up relationships between entities. You will use this functionality later in the mod!
+
+<!-- 
+This syntax is a bit different from what we are used to. We use `attribute` singular, and then as a symbol we pick the name of what we want our attribute to be. We use a do end block similar to an enumerable with a block parameter. Now the block parameter, `object` is a lot like self. We get to use it for each single thing of a collection we pass to the serializer. We are essentially saying for each thing you serialize, grab the books and count them too. In this manner we can add a custom generated value for each book.
```

Comment:
Trying to think if there is something a bit more spicy we could add here to introduce this topic without having model associations. Any ideas?